garage/replay_buffer/her_replay_buffer.py

```diff
@@ -0,0 +1,170 @@
+
+See: https://arxiv.org/abs/1707.01495.
+"""
+import inspect
+from typing import Callable
+
+import numpy as np
+
+from garage.misc.overrides import overrides
+
+
+def make_her_sample(replay_k: int, reward_fun: Callable):
+    """
+    Generate a transition sampler for HER ReplayBuffer.
+
+    :param replay_k: the ratio between HER replays and regular replays
+    :param reward_fun: function to re-compute the reward with substituted goals
+    :return:
+    """
+    future_p = 1 - (1. / (1 + replay_k))
+
+    def _her_sample_transitions(episode_batch, sample_batch_size):
+        """
+        Generate a dictionary of transitions.
+
+        :param episode_batch: [batch_size, T, dim]
+        :param sample_batch_size: batch_size per sample.
+        :return: transitions which transitions[key] has the shape of
+        [sample_batch_size, dim].
+        """
+        # Select which episodes to use
+        time_horizon = episode_batch["action"].shape[1]
+        rollout_batch_size = episode_batch["action"].shape[0]
+        episode_idxs = np.random.randint(0, rollout_batch_size,
+                                         sample_batch_size)
+        # Select time steps to use
+        t_samples = np.random.randint(time_horizon, size=sample_batch_size)
+        transitions = {
+            key: episode_batch[key][episode_idxs, t_samples].copy()
+            for key in episode_batch.keys()
+        }
+
+        her_indexes = np.where(
+            np.random.uniform(size=sample_batch_size) < future_p)
+        future_offset = np.random.uniform(size=sample_batch_size) * (
+            time_horizon - t_samples)
+        future_offset = future_offset.astype(int)
+        future_t = (t_samples + 1 + future_offset)[her_indexes]
+
+        future_ag = episode_batch["achieved_goal"][episode_idxs[her_indexes],
+                                                   future_t]
+        transitions["goal"][her_indexes] = future_ag
+
+        # Reconstruct info dictionary for reward computation.
+        info = {}
+        for key, value in transitions.items():
+            if key.startswith("info_"):
+                info[key.replace("info_", "")] = value
+
+        # Re-compute reward since we may have substituted the goal.
+        reward_params_keys = inspect.signature(reward_fun).parameters.keys()
+        reward_params = {
+            rk: transitions[k]
+            for k, rk in zip(["next_achieved_goal", "goal"],
+                             list(reward_params_keys)[:-1])
+        }
+        reward_params["info"] = info
+        transitions["reward"] = reward_fun(**reward_params)
+
+        transitions = {
+            k: transitions[k].reshape(sample_batch_size,
+                                      *transitions[k].shape[1:])
+            for k in transitions.keys()
+        }
+        assert (transitions["action"].shape[0] == sample_batch_size)
+        return transitions
+
+    return _her_sample_transitions
+
+
+class HerReplayBuffer(object):
```

Comment:
Python 3 classes don't need to inherit from object:

`class HerReplayBuffer:`

Reply:
Done.

Reply:
Done.